src/main/java/org/ice4j/ice/ComponentSocket.java

```diff
@@ -0,0 +1,206 @@
+/*
+ * ice4j, the OpenSource Java Solution for NAT and Firewall Traversal.
+ *
+ * Copyright @ 2015-2016 Atlassian Pty Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ice4j.ice;
+
+import org.ice4j.*;
+import org.ice4j.socket.*;
+
+import java.beans.*;
+import java.net.*;
+import java.util.*;
+import java.util.concurrent.*;
+import org.ice4j.util.*; //Disambiguation
+
+/**
+ * Extends {@link MergingDatagramSocket} with functionality specific to
+ * an ICE {@link Component}.
+ */
+public class ComponentSocket
+    extends MergingDatagramSocket
+    implements PropertyChangeListener
+{
+    /**
+     * The {@link Logger} used by the {@link MergingDatagramSocket} class and
+     * its instances for logging output.
+     */
+    private static final java.util.logging.Logger classLogger
+        = java.util.logging.Logger.getLogger
+                (MergingDatagramSocket.class.getName());
+
+    /**
+     * Controls access to {@link #authorizedAddresses}.
+     */
+    private final Object authorizedAddressesSyncRoot = new Object();
+
+    /**
+     * The set of remote addresses, which this socket is allowed to receive
+     * from. These should be the addresses which we have confirmed (e.g. by
+     * having received a STUN message with correct authentication fields).
+     */
+    private Set<SocketAddress> authorizedAddresses
+        = Collections.newSetFromMap(
+                new ConcurrentHashMap<SocketAddress, Boolean>());
```

Comment:
i'm a bit confused by this...why back this set with a concurrent hashmap?  also, why back it with a map at all?  we create the map in place here, so there's no accessing it directly.  why not just create a set?  if the purpose of backing it by the concurrenthashmap is to inherit the thread safety, it seems like we don't get to take full advantage of that anyway as half the accesses are in a synchronized block (although i'm sure 'accept' is called much more frequently, so maybe this is all just to optimize for that path?)